Neos.ContentRepository.Core/Classes/EventStore/EventPersister.php

```diff
@@ -42,17 +47,43 @@ public function publishEvents(ContentRepository $contentRepository, EventsToPubl
         $normalizedEvents = Events::fromArray(
             $eventsToPublish->events->map(fn(EventInterface|DecoratedEvent $event) => $this->normalizeEvent($event))
         );
-        $this->eventStore->commit(
+        $commitResult = $this->eventStore->commit(
             $eventsToPublish->streamName,
             $normalizedEvents,
             $eventsToPublish->expectedVersion
         );
+        $hooks = $this->hooksFactory->build($contentRepository);
+        $store = new SemaphoreStore();
+        $factory = new LockFactory($store);
+        $lock = $factory->createLock($contentRepository->id->value . '_catchup');
+        $lock->acquire(true);
+
+        $expectedCheckpoint = SequenceNumber::fromInteger($commitResult->highestCommittedSequenceNumber->value - $eventsToPublish->events->count());
+
+        $projectionsToUpdate = [];
         foreach ($this->projections as $projection) {
+            if (!$projection->getCheckpoint()->equals($expectedCheckpoint)) {
+                //throw new \RuntimeException(sprintf('Projection %s is at checkpoint %d, but was expected to be at %d', $projection::class, $projection->getCheckpoint()->value, $expectedCheckpoint->value), 1714062281);
+                continue;
+            }
+            $projectionsToUpdate[] = $projection;
             if ($projection instanceof WithMarkStaleInterface) {
                 $projection->markStale();
             }
         }
-        $contentRepository->catchUpProjections();
+
+        $hooks->dispatchBeforeCatchUp();
+        foreach ($this->eventStore->load($eventsToPublish->streamName)->withMinimumSequenceNumber($expectedCheckpoint->next()) as $eventEnvelope) {
+            $event = $this->eventNormalizer->denormalize($eventEnvelope->event);
+            $hooks->dispatchBeforeEvent($event, $eventEnvelope);
+            foreach ($projectionsToUpdate as $projection) {
+                $projection->apply($event, $eventEnvelope);
```

Comment:
TODO: `try/catch` -> collect exceptions and throw one combined exception after releasing the lock in line 85